File: in_a_Weekend/python/material.py
import math
from random import random

from vec3 import Vec3
from ray import Ray

class ScatterRecord(object):

    def __init__(self, attenuation, scattered):
        self.attenuation = attenuation
        self.scattered = scattered

class Material(object):

    def scatter(self, r_in, hit):
        raise NotImplementedError

def random_in_unit_sphere():
    while True:
        p = 2.0*Vec3(random(),random(),random()) - Vec3(1,1,1)
        if p.dot(p) < 1: return p

class Lambertian(Material):

    def __init__(self, a):
        self.albedo = a

    def scatter(self, r_in, hit):
        target = hit.p + hit.normal + random_in_unit_sphere()
        return ScatterRecord(self.albedo, Ray(hit.p, target-hit.p))

def reflect(v, n):
    return v - 2*v.dot(n)*n

class Metal(Material):

    def __init__(self, a, f=0.3):
        self.albedo = a
        self.fuzz = min(f, 1.0)

    def scatter(self, r_in, hit):
        reflected = reflect(r_in.direction.unit, hit.normal)
        scattered = Ray(hit.p, reflected + self.fuzz*random_in_unit_sphere())
        if scattered.direction.dot(hit.normal) > 0:
            return ScatterRecord(self.albedo, scattered)
        return False

def refract(v, n, ni_over_nt):
    uv = v.unit
    dt = uv.dot(n)
    discriminant = 1.0 - ni_over_nt*ni_over_nt*(1-dt*dt)
    if discriminant > 0:
        return ni_over_nt*(uv - n*dt) - n*math.sqrt(discriminant)
    return False

def schlick(cosine, ref_idx):
    r0 = (1-ref_idx) / (1+ref_idx)
    r0 *= r0
    return r0 + (1-r0)*math.pow((1-cosine), 5)

class Dielectric(Material):

    def __init__(self, ri):
        self.ref_idx = ri

    def scatter(self, r_in, hit):
        reflected = reflect(r_in.direction, hit.normal)
        attenuation = Vec3(1.0, 1.0, 1.0)

        if r_in.direction.dot(hit.normal) > 0:
            outward_normal = -hit.normal
            ni_over_nt = self.ref_idx
            cosine = self.ref_idx * r_in.direction.dot(hit.normal) / r_in.direction.length
        else:
            outward_normal = hit.normal
            ni_over_nt = 1.0 / self.ref_idx
            cosine = -r_in.direction.dot(hit.normal) / r_in.direction.length

        refracted = refract(r_in.direction, outward_normal, ni_over_nt)
        if not refracted:
            reflect_prob = 1.0
        else:
            reflect_prob = schlick(cosine, self.ref_idx)

        if random() < reflect_prob:
            return ScatterRecord(attenuation, Ray(hit.p, reflected))

        return ScatterRecord(attenuation, Ray(hit.p, refracted))

File: in_a_Weekend/rust/rti1w/src/material.rs
extern crate rand;

use rand::Rng;

use vec3::Vec3;
use ray::Ray;
use hitable::HitRecord;

pub struct ScatterRecord {
    pub attenuation: Vec3,
    pub scattered: Ray
}

/*************************************************************************************************/
pub trait Material {
    fn scatter(&self, r_in: &Ray, hit: &HitRecord) -> Option<ScatterRecord>;
}

/*************************************************************************************************/
pub struct Lambertian {
    albedo: Vec3
}

impl Lambertian {
    pub fn new(albedo: Vec3) -> Lambertian {
        Lambertian { albedo }
    }
}

fn random_in_unit_sphere() -> Vec3 {
    let mut rng = rand::thread_rng();
    let mut r = || -> f32 { rng.gen() };
    loop {
        let p = 2.0*Vec3::new(r(), r(), r()) - Vec3::new(1.,1.,1.);
        if p.dot(p) < 1.0 { return p }
    }
}

impl Material for Lambertian {
    fn scatter(&self, _r_in: &Ray, hit: &HitRecord) -> Option<ScatterRecord> {
        let target = hit.p + hit.normal + random_in_unit_sphere();
        return Some(ScatterRecord {
            attenuation: self.albedo,
            scattered: Ray::new(hit.p, target-hit.p)
        })
    }
}

/*************************************************************************************************/
pub struct Metal {
    albedo: Vec3,
    fuzz: f32
}

impl Metal {
    pub fn new(albedo: Vec3, fuzz: f32) -> Metal {
        Metal { albedo, fuzz: fuzz.min(1.) }
    }
}

fn reflect(v: Vec3, n: Vec3) -> Vec3 {
    v - 2.0*v.dot(n)*n
}

impl Material for Metal {
    fn scatter(&self, r_in: &Ray, hit: &HitRecord) -> Option<ScatterRecord> {
        let reflected = reflect(Vec3::unit(r_in.direction), hit.normal);
        let scattered = Ray::new(hit.p, reflected + self.fuzz*random_in_unit_sphere());
        if scattered.direction.dot(hit.normal) > 0. {
            Some(ScatterRecord {
                attenuation: self.albedo,
                scattered
            })
        } else {
            None
        }
    }
}

/*************************************************************************************************/
fn refrect(v: Vec3, n: Vec3, ni_over_nt: f32) -> Option<Vec3> {
    let uv = Vec3::unit(v);
    let dt = uv.dot(n);
    let discriminant = 1.0 - ni_over_nt*ni_over_nt*(1.0-dt*dt);
    if discriminant > 0.0 {
        Some(ni_over_nt*(uv - n*dt) - n*discriminant.sqrt())
    } else {
        None
    }
}

fn schlick(cosine: f32, ref_idx: f32) -> f32 {
    let mut r0 = (1.0-ref_idx) / (1.0+ref_idx);
    r0 *= r0;
    return r0 + (1.0-r0)*(1.0-cosine).powf(5.0);
}

pub struct Dielectric {
    pub ref_idx: f32
}

impl Dielectric {
    pub fn new(ref_idx: f32) -> Dielectric {
        Dielectric { ref_idx }
    }
}

impl Material for Dielectric {
    fn scatter(&self, r_in: &Ray, hit: &HitRecord) -> Option<ScatterRecord> {
        let reflected = reflect(r_in.direction, hit.normal);
        let attenuation = Vec3::new(1.0, 1.0, 1.0);
        let outward_normal;
        let ni_over_nt;
        let reflect_prob;
        let cosine;
        if r_in.direction.dot(hit.normal) > 0.0 {
            outward_normal = -hit.normal;
            ni_over_nt = self.ref_idx;
            cosine = self.ref_idx * r_in.direction.dot(hit.normal) / r_in.direction.length();
        } else {
            outward_normal = hit.normal;
            ni_over_nt = 1.0 / self.ref_idx;
            cosine = -r_in.direction.dot(hit.normal) / r_in.direction.length();
        }
        if let Some(refracted) = refrect(r_in.direction, outward_normal, ni_over_nt) {
            reflect_prob = schlick(cosine, self.ref_idx);
            if rand::random::<f32>() < reflect_prob {
                Some(ScatterRecord { attenuation, scattered: Ray::new(hit.p, reflected) })
            } else {
                Some(ScatterRecord { attenuation, scattered: Ray::new(hit.p, refracted) })
            }
        } else {
            Some(ScatterRecord { attenuation, scattered: Ray::new(hit.p, reflected) })
        }
    }
}
